### PR TITLE
[Go] [Abstractions] Untyped Nodes

### PR DESCRIPTION
**Pull request by andrueastman**
This PR is related to https://github.com/microsoft/kiota/issues/2319

It adds a number of types that implement the Parsable interface to enable the representation of types that are unknown at generation time to unblock the serialization and deserialization of properties of unknown types. 

Generation of these types is unblocked in https://github.com/microsoft/kiota/pull/4095

Partially resolves https://github.com/microsoftgraph/msgraph-sdk-go/issues/473

### CODE REVIEW

File: serialization/untyped_boolean.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+// GetValue returns the bool value.
+func (un *UntypedBoolean) GetValue() *bool {
+	return un.value.(*bool)
+}
+
+// NewUntypedBoolean creates a new UntypedBoolean object.
+func NewUntypedBoolean(boolValue *bool) *UntypedBoolean {
+	m := &UntypedBoolean{}
+	m.value = boolValue
```
Comment (rkodev) [inline serialization/untyped_boolean.go:16]: Its possible to assign a `nil` as the `boolValue` possibly constructing an untyped boolean thats an untyped null.

Comment (andrueastman) [reply serialization/untyped_boolean.go:16]: Thanks for taking a look @rkodev.
Just to confirm here, what you're saying is that you can pass a nil as below?

```golang
fakeBoolValue := NewUntypedBoolean(nil)
```

And the suggestion would be to add code that guards from such scenarios?

Comment (rkodev) [reply serialization/untyped_boolean.go:16]: Yes, we should throw an error when a nil value is passed for as the argument for a constructor

File: serialization/untyped_boolean.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	UntypedNode
+}
+
+// GetValue returns the bool value.
+func (un *UntypedBoolean) GetValue() *bool {
+	return un.value.(*bool)
+}
+
+// NewUntypedBoolean creates a new UntypedBoolean object.
+func NewUntypedBoolean(boolValue *bool) (*UntypedBoolean, error) {
```
Comment (baywet) [inline serialization/untyped_boolean.go:16]: we're probably going to get feedback along the line of this being really convoluted in GO:
1. accept a reference to a bool
2. if the reference is nil, return an error

We should probably accept the boolean value directly, unless there's another constraint I'm not seeing?
And then we can make a reference to that value in the body to facilitate the later type assertion in the getter.
This way the constructor doesn't need to return an error anymore.

File: serialization/untyped_boolean.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import "errors"
+
+// UntypedBoolean defines an untyped boolean object.
+type UntypedBoolean struct {
+	UntypedNode
+}
+
+// GetValue returns the bool value.
+func (un *UntypedBoolean) GetValue() *bool {
+	return un.value.(*bool)
```
Comment (baywet) [inline serialization/untyped_boolean.go:12]: this is going to cause a panic
```suggestion
    castValue, ok := un.value.(*bool)
    if ok {
	   return castValue
	}
	return nil
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: serialization/untyped_boolean.go (hunk 2 above)
Comment (andrueastman) [reply serialization/untyped_boolean.go]: Thanks for the pointer here. Really makes things clearer.

**Pull request merged**